### PR TITLE
msq: add multi-stage-query docs

### PR DESCRIPTION
**Pull request by 317brian**
This PR adds all the documentation updates for the multi-stage query architecture and the MSQ task engine

Note that the screenshot on the Druid console page is outdated and will get updated in a subsequent PR.

This PR has:
- [x] been self-reviewed.

### CODE REVIEW

**Comment by 317brian** [comment]
@2bethere @vogievetsky please review the OSS docs as a whole for MSQ

@cryptoe please review the known issues list to see if there's anything to be added or removed: https://github.com/apache/druid/pull/12983/files#diff-e83f8a116cd5e34642cfe3474e915dd586d4679547ac393a6dea654a78c9bbec

File: docs/multi-stage-query/index.md [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+If you use `PARTITIONED BY ALL` or `PARTITIONED BY ALL TIME`, time-based
+partitioning is disabled. In these cases, your ingestion query doesn't need
+to include a `__time` column. However, Druid still creates a `__time` column 
+in your Druid table and sets all timestamps to 1970-01-01 00:00:00.
+
+For more information, see [Primary timestamp](../ingestion/data-model.md#primary-timestamp).
+
+### PARTITIONED BY
+
+INSERT and REPLACE queries require the PARTITIONED BY clause, which determines how time-based partitioning is done. In Druid, data is split into segments, one or more per time chunk defined by the PARTITIONED BY granularity. A good general rule is to adjust the granularity so that each segment contains about five million rows. Choose a granularity based on your ingestion rate. For example, if you ingest a million rows per day, PARTITION BY DAY is good. If you ingest a million rows an hour, choose PARTITION BY HOUR instead.
```
Comment (cryptoe) [inline docs/multi-stage-query/index.md:181]: Should we make 5 million bold ?

File: docs/multi-stage-query/msq-known-issues.md [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+- INSERT with column lists, like
+  `INSERT INTO tbl (a, b, c) SELECT ...`, is not implemented.
+
+## EXTERN queries
+
+- EXTERN does not accept `druid` input sources.
+
+## Missing guardrails
+
```
Comment (cryptoe) [inline docs/multi-stage-query/msq-known-issues.md:101]: Maximum number of input files. No guardrail today means the controller can potentially run out of memory tracking them all.

**Review by cryptoe — APPROVED** [review]
LGTM

File: docs/multi-stage-query/msq-tutorial-connect-external-data.md [hunk 3]
```diff
@@ -0,0 +1,126 @@
+---
+id: connect-external-data
+title: Tutorial - Connect external data
+description: How to generate a query that references externally hosted data
+---
+
+> The SQL-based ingestion using the multi-stage query task engine is our recommended solution starting in Druid 24.0. Alternative ingestion solutions, such as native batch and Hadoop-based ingestion systems, will still be supported. We recommend you read all [known issues](./msq-known-issues.md) and test the feature in a development environment before rolling it out in production. Using the multi-stage query task engine with `SELECT` statements that do not write to a datasource is experimental.
```
Comment (317brian) [inline docs/multi-stage-query/msq-tutorial-connect-external-data.md:7]:
```suggestion
> SQL-based ingestion using the multi-stage query task engine is our recommended solution starting in Druid 24.0. Alternative ingestion solutions, such as native batch and Hadoop-based ingestion systems, will still be supported. We recommend you read all [known issues](./msq-known-issues.md) and test the feature in a development environment before rolling it out in production. Using the multi-stage query task engine with `SELECT` statements that do not write to a datasource is experimental.
```

File: docs/multi-stage-query/msq-durable-storage.md [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+---
+
+> SQL-based ingestion using the multi-stage query task engine is our recommended solution starting in Druid 24.0. Alternative ingestion solutions, such as native batch and Hadoop-based ingestion systems, will still be supported. We recommend you read all [known issues](./msq-known-issues.md) and test the feature in a development environment before rolling it out in production. Using the multi-stage query task engine with `SELECT` statements that do not write to a datasource is experimental.
+
+By default, the multi-stage query task engine (MSQ task engine) uses the local storage of a node to store data from intermediate steps when executing a query. Although this method provides better speed when executing a query, the data is lost if the node encounters an issue. When you enable durable storage, intermediate data is stored in Amazon S3 instead. Using this feature can improve the reliability of queries that use more than 20 workers. In essence, you trade some performance for better reliability. This is especially useful for long-running queries.
```
Comment (vogievetsky) [inline docs/multi-stage-query/msq-durable-storage.md:10]: Where does the "20 workers" language come from?

File: docs/multi-stage-query/msq-reference.md [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+```
+
+EXTERN consists of the following parts:
+
+1.  Any [Druid input source](../ingestion/native-batch-input-source.md) as a JSON-encoded string.
+2.  Any [Druid input format](../ingestion/data-formats.md) as a JSON-encoded string.
+3.  A row signature, as a JSON-encoded array of column descriptors. Each column descriptor must have a `name` and a `type`. The type can be `string`, `long`, `double`, or `float`. This row signature is used to map the external data into the SQL layer.
+
+### INSERT
+
```
Comment (vogievetsky) [inline docs/multi-stage-query/msq-reference.md:82]: I think somewhere here we should call out a note about how the MSQ INSERT syntax (REPLACE also, but this has more impact for INSERT) deviates from the SQL standard in that the column are mapped by name and not positionally.

Maybe something like:

```
Please note that unlike standard SQL the data is inserted according to column name and not positionally which means that it is important to get the output column names of subsequent inserts to be the same as the table and not to simply reply on their positions within the SELECT clause.
```

File: docs/multi-stage-query/msq-api.md [hunk 6]
```diff
@@ -0,0 +1,1634 @@
...
+  "taskId": "query-f795a235-4dc7-4fef-abac-3ae3f9686b79",
+  "state": "RUNNING",
+}
+```
+
+**Response fields**
+
+|Field|Description|
+|-----|-----------|
+ | taskId | Controller task ID. You can use Druid's standard [task APIs](../operations/api-reference.md#overlord) to interact with this controller task.|
```
Comment (vogievetsky) [inline docs/multi-stage-query/msq-api.md:104]: is there a strange indent here or is the GitHub UI buggy

File: docs/multi-stage-query/msq-api.md [hunk 7]
```diff
@@ -0,0 +1,1634 @@
...
+
+## Submit a query
+
+You submit queries to the MSQ task engine using the `POST /druid/v2/sql/task/` endpoint.
+
+### Request
+
+Currently, the MSQ task engine ignores the provided values of `resultFormat`, `header`,
+`typesHeader`, and `sqlTypesHeader`. SQL SELECT queries always behave as if `resultFormat` is an array, `header` is
+true, `typesHeader` is true, and `sqlTypesHeader` is true.
```
Comment (vogievetsky) [inline docs/multi-stage-query/msq-api.md:24]: I would change

```
SQL SELECT queries always behave as if `resultFormat` is an array, `header` is
true, `typesHeader` is true, and `sqlTypesHeader` is true.
```

To

```
SQL SELECT queries write out their results into the task report (in the `multiStageQuery.payload.results.results` key) formatted as if `resultFormat` is an `array`.
```

**Review by techdocsmith — APPROVED** [review]
LGTM

File: docs/multi-stage-query/msq-security.md [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+  - INSERT or REPLACE into a Druid datasource requires the WRITE DATASOURCE permission on that
+  datasource.
+  - EXTERN references to external data require READ permission on the resource name "EXTERNAL" of the resource type "EXTERNAL". Users without the correct permission encounter a 403 error when trying to run queries that include EXTERN.
+
+Query tasks that you submit to the MSQ task engine are Overlord tasks, so they follow the Overlord's (indexer) model. This means that users with access to the Overlord API can perform some actions even if they didn't submit the query. The actions include retrieving the status or canceling a query. For more information about the Overlord API and the task API, see [APIs for SQL-based ingestion](./msq-api.md).
+
+To interact with a query through the Overlord API, you need the following permissions:
+
+- INSERT or REPLACE queries: You must have READ DATASOURCE permission on the output datasource.
+- SELECT queries: You must have read permissions on the `__query_select` datasource, which is a stub datasource that gets created.
```
Comment (vogievetsky) [inline docs/multi-stage-query/msq-security.md:24]: This should read `... that does not get created.`

File: website/sidebars.json [hunk 9]
```diff
@@ -66,6 +68,17 @@
       "ingestion/tasks",
       "ingestion/faq"
     ],
+  "SQL-based ingestion": [
```
Comment (vogievetsky) [inline website/sidebars.json:71]: this needs two spaces of leading indent

**Comment by 317brian** [comment]
The CI is failing on the licensing check and JDK 8 packaging check: 
image

**Comment by abhishekagarwal87** [comment]
@317brian - You need to add the copyright header in text files.

**Review by vogievetsky — APPROVED** [review]

**Pull request merged**